pmlc/dialect/pxa/transforms/stencil_generic.cc

```diff
@@ -0,0 +1,164 @@
+    const llvm::SmallVector<mlir::BlockArgument, 8> &bound_idxs,
+    const llvm::SmallVector<mlir::Value, 3> &tensors) {
+  auto curr_idx = bound_idxs.size();
+  if (curr_idx == semantic_idx_count) {
+    // This is a legal binding, append it and we're done with this branch
+    legal_permutations.push_back(
+        TensorAndIndexPermutation(tensors, bound_idxs));
+  } else {
+    for (const auto &block_arg : block_args) {
+      // Don't bind same index twice
+      if (std::find(bound_idxs.begin(), bound_idxs.end(), block_arg) !=
+          bound_idxs.end()) {
+        continue;
+      }
+
+      // Verify the requirements for this index with each tensor are all met
+      bool reqs_met = true;
+      for (size_t i = 0; i < tensors.size(); i++) {
+        try { // TODO: probably don't keep long term
+          if (!requirements.at(std::make_pair(i, curr_idx))(tensors[i],
+                                                            block_arg)) {
+            reqs_met = false;
+            break;
+          }
+        } catch (const std::out_of_range &e) {
+          IVLOG(1, "Error message: " << e.what());
+          IVLOG(1, "Requested key was: " << std::make_pair(i, curr_idx));
+          throw;
+        }
+      }
+      if (!reqs_met) {
+        continue;
+      }
+
+      // If we made it to here, this index has appropriate semantics; bind it
+      // and recurse
+      llvm::SmallVector<mlir::BlockArgument, 8> new_bound_idxs(bound_idxs);
+      new_bound_idxs.push_back(block_arg);
+      RecursiveBindIndex(new_bound_idxs, tensors);
+    }
+  }
+}
+
+void StencilGeneric::DoStenciling() {
+  // Initialization
+  // TODO: Something like?: tensors.clear();
+  {
+    auto maybe_loads_and_stores = capture();
+    if (maybe_loads_and_stores) {
+      loads_and_stores = maybe_loads_and_stores.getValue();
+    } else {
+      IVLOG(4, "Cannot Stencil: Operations fail to pattern-match.");
+      return;
+    }
+  }
+
+  llvm::SmallVector<mlir::Value, 3> tensors;
+  for (auto &load_op : loads_and_stores.loads) {
+    // TODO: It's probably fine, but I'm a little worried `getMemRef` isn't
+    // const
+    tensors.push_back(load_op.getMemRef());
+  }
+  size_t first_store_idx = tensors.size();
+  for (auto &store_op : loads_and_stores.stores) {
+    // TODO: It's probably fine, but I'm a little worried `out` isn't const
+    tensors.push_back(store_op.out());
+    // TODO: Probably should handle reduces vs. true stores in a different way
+    // if (auto reduce_op = llvm::dyn_cast_or_null<AffineReduceOp>(store_op)) {
+    //   tensors.push_back(reduce_op.out());
+    // } else if (auto true_store_op =
+    // llvm::dyn_cast_or_null<mlir::AffineStoreOp>(store_op)) {
+    //   tensors.push_back(true_store_op.getMemRef());
+    // } else {
+    //   // TODO: throw?
+    //   IVLOG(1, "Unexpected failure to load tensors from ops in stenciling");
+    //   return;
+    // }
+  }
+  std::sort(tensors.begin(), tensors.begin() + first_store_idx);
```

Comment:
Is there a reason not to store the input and output tensors in separate vectors to avoid some of this complexity?

Reply:
No. I think I was trying to reduce the number of parameters on `RecursiveBindIndex`? But that's not actually a good reason.

Reply:
Actually doing the related refactor, it is doing something a little nontrivial: either we need separate `requirements` for loads and stores, or we need to at some point join the loads & stores into one vector. While perhaps this is premature optimization, it feels better to do that out here than in the guts of the recursive function -- especially since the complexity has to go _somewhere_.

For the moment, I'm just going to make this look a little cleaner with a iterator to the split location.

Reply:
Fair point, but another way to go is to use ArrayRef.  Basically, you can do:

SmallVector<Value, 3> all = <something>;
auto inputs = ArrayRef<Value>(x).slice(0, num_inputs);
auto outputs =  ArrayRef<Value>(x).slice(num_inputs, num_outputs);

for (Value x : inputs) {
  <goo>
}

or:

Value v = outputs[i];  // Acesses all[num_inputs + i]

or whatever.  So it's easy to make a 'slice' of a vector that works like a full vector for all std calls, etc.

Reply:
It appears `ArrayRef`s don't have `swap`, which precludes the necessary `std::sort` (and presumably also `std::next_permutation`)